### ISSUE TITLE
ctx.req.user being called by a dependency

### ISSUE THREAD

**Issue by rkusa** [rkusa/koa-passport#66]
Issue will be tracked here

@jeffijoe in #57
> It appears that the `verify` callback in strategies use the req setter. E.g. https://github.com/jaredhanson/passport-http/blob/master/lib/passport-http/strategies/basic.js#L95
> 
> I checked the stack trace:
> 
> ```
> at IncomingMessage.Object.defineProperty.set [as user] (\node_modules\koa-passport\lib\framework\koa.js:37:19)
> at Object.properties.(anonymous function).set (\node_modules\koa-passport\lib\framework\request.js:113:16)
> at Object.req.login.req.logIn (\node_modules\passport\lib\http\request.js:44:18)
> at Object.<anonymous> (\node_modules\koa-passport\lib\framework\request.js:104:27)
> at BasicStrategy.strategy.success (\node_modules\passport\lib\middleware\authenticate.js:235:13)
> at verified (\node_modules\passport-http\lib\passport-http\strategies\basic.js:91:10)
> ```
> 
> So this deprecation notice is unavoidable it seems.
> 
> Should I open a new issue?

@mapmeld in rkusa/koa-passport-example#20

> With the latest update to koa-passport, I get a ```koa-passport deprecated setting ctx.req.user is deprecated in favour of ctx.state.user``` warning when I log in
> 
> This codebase doesn't use ```ctx.req``` or ```ctx.state```, so my best guess is that it's coming from ```koa-route``` or another dependency.

**Comment by rkusa** [comment]
I've removed the deprecation warning for now; published as `2.2.2`. The `user` is still accessible through `ctx.state.user`. @gingerich I am still eager to adhere to Koa's way of doing things. However, I've test implemented workarounds to get passport to work with having the `user` in `ctx.state`. But since methods like `login()` are actively using `req.user`, the amount of required workarounds is unreasonable high. Maybe an ideal solution can only be achieved by implementing authentication for Koa from scratch. Unfortunately, this is not possible with keeping compatibility to all passport strategies, which would make such an undertaking a tremendous and unreasonable amount of work 😕

**Comment by jeffijoe** [comment]
@rkusa not guessing monkey-patching all `use`d strategies would work?

**Comment by rkusa** [comment]
Thanks for the feedback! Since passport is [adding things to `req` anyway](https://github.com/jaredhanson/passport/blob/master/lib/middleware/authenticate.js#L82:L85), the question would be: what is better, monkey-patching all strategies, or living with having the `user` property in multiple places. I personally don't like both, what probably prefer the latter. But I am open to hearing other opinions on that!

**Comment by jeffijoe** [comment]
@rkusa the best option would be making a different authentication library (clean codebase), but it would be a shame to discard the sheer amount of Passport strategies available.

**Comment by jeffijoe** [comment]
Easiest path would be proxying the `ctx.req.user` getters and setters to `ctx.state` - that way everyone is happy I guess. :shipit:

**Comment by rkusa** [comment]
Yep, in an ideal world we would have a passport written for Koa from scratch. But writing all the strategies again... and going through all security issues again... does not sound too good.

I am currently copying the user value over to `ctx.state`. A getter and setter would definitely better. Especially since the user on the `req` is already a getter and setter. However, I ran into a weird bug tomorrow and I hadn't the time to debug it, which is why I went for copying for now. I'll replace this by a getter and setter later.
